lib/charms/prometheus_k8s/v0/prometheus.py

```diff
@@ -897,3 +895,262 @@ def _scrape_metadata(self) -> dict:
+
+            relation.data[self._charm.app]["scrape_jobs"] = json.dumps(jobs)
+
+    def _update_alert_rules(self, event):
+        if not (unit_rules := self._get_alert_rules(event.relation)):
+            return
+
+        appname = event.relation.app.name
+        rules = self._label_alert_rules(unit_rules, appname)
+        updated_group = {"name": self._group_name(appname), "rules": rules}
+
+        for relation in self.model.relations[self._prometheus_relation]:
+            alert_rules = json.loads(relation.data[self._charm.app].get("alert_rules", "{}"))
+            groups = alert_rules.get("groups", [])
+            groups = [group for group in groups if updated_group["name"] != group["name"]]
+            groups.append(updated_group)
+            relation.data[self._charm.app]["alert_rules"] = json.dumps({"groups": groups})
+
+    def _remove_alert_rules(self, event):
+        group_name = self._group_name(event.relation.app.name)
+        unit_name = event.unit.name
+
+        for relation in self.model.relations[self._prometheus_relation]:
+            if not (
+                alert_rules := json.loads(relation.data[self._charm.app].get("alert_rules", "{}"))
+            ):
+                continue
+
+            if not (groups := alert_rules.get("groups", [])):
+                continue
+
+            if not (
+                changed_group := list(filter(lambda group: group["name"] == group_name, groups))
+            ):
+                continue
+            changed_group = changed_group[0]
+
+            groups = list(filter(lambda group: group["name"] != group_name, groups))
+
+            rules_kept = list(
+                filter(
+                    lambda rule: rule.get("labels").get("juju_unit") != unit_name,
+                    changed_group.get("rules"),
+                )
+            )
+
+            if rules_kept:
+                changed_group["rules"] = rules_kept
+                groups.append(changed_group)
+
+            relation.data[self._charm.app]["alert_rules"] = (
+                json.dumps({"groups": groups}) if groups else "{}"
+            )
+
+    def _get_targets(self, relation):
+        targets = {}
+        for unit in relation.units:
+            port = relation.data[unit].get("port", 80)
+            if hostname := relation.data[unit].get("hostname"):
+                targets.update({unit.name: {"hostname": hostname, "port": port}})
+
+        return targets
+
+    def _get_alert_rules(self, relation):
+        rules = {}
+        for unit in relation.units:
+            if unit_rules := yaml.safe_load(relation.data[unit].get("groups", "")):
+                rules.update({unit.name: unit_rules})
+
+        return rules
+
+    def _job_name(self, appname):
+        return f"juju_{self.model.name}_{self.model.uuid[:7]}_{appname}_prometheus_scrape"
+
+    def _group_name(self, appname):
+        return f"juju_{self.model.name}_{self.model.uuid[:7]}_{appname}_alert_rules"
```

Comment:
If you implement the above, this could be: `f"{self.app_id}_alert_rules"`

Reply:
same comment as for `_job_name`.